### PR TITLE
Feature: Get Nested Agents in a `GroupChat`

### LINKED ISSUE #1632: [Feature Request]: Listing out all agents in a GroupChatManager
### Is your feature request related to a problem? Please describe.

I'm not sure if this is currently possible, but it would be useful to list out all the agents present in a `GroupChatManager`. 

### Describe the solution you'd like

I was thinking of something similar to `agent_by_name` in `GroupChat`
```python
def agent_by_name(self, name: str) -> Optional[Agent]:
        """Returns the agent with a given name."""
        return self.present_agents().get(name)

def present_agents(self) -> Dict[str, Agent]:
        agents = dict()
        for agent in self._groupchat.agents:
            if not isinstance(agent, GroupChatManager):
                agents[agent.name] = agent
            else:
                agents.update(agent.present_agents())
        return agents
```
I haven't tested this implementation but should be close.

### Additional context

This could be useful when you're defining teams in different modules

--- file_a
```python
def some_team():
    member1 = AssistantAgent(...)
    member2 = UserProxyAgent(...)

    gc = GroupChat(agents=[member1, member2], ...)
    return GroupChatManager(groupchat=gc, ...)
```

--- file_b
```python
import file_a

team1 = file_a.some_team()

# I need access to the agents in team1
agents = team1.present_agents()
```

### PR DESCRIPTION
**Pull request by WaelKarkoub**
## Why are these changes needed?

Adds the capability to get all the agents in a given a `GroupChatManager`. Improves quality of life for the following use case:

--- file_a
```python
def some_team():
    member1 = AssistantAgent(...)
    member2 = UserProxyAgent(...)

    gc = GroupChat(agents=[member1, member2], ...)
    return GroupChatManager(groupchat=gc, ...)
```

--- file_b
```python
import file_a

team1 = file_a.some_team()

# I need access to the agents in team1
agents = team1.all_agents()
```
[Example use case.](https://gist.github.com/WaelKarkoub/9b5d8324d6abc334efcc4d2b5ba5ae66)

## Related issue number

Closes #1632 

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by WaelKarkoub** [comment]
> @WaelKarkoub please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement
> ### Contribution License Agreement
> This Contribution License Agreement (**“Agreement”**) is agreed to by the party signing below (**“You”**), and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your contributions to Microsoft open source projects. This Agreement is effective as of the latest signature date below.
> 
> 1. **Definitions**.
>    **“Code”** means the computer software code, whether in human-readable or machine-executable form,
>    that is delivered by You to Microsoft under this Agreement.
>    **“Project”** means any of the projects owned or managed by Microsoft and offered under a license
>    approved by the Open Source Initiative ([www.opensource.org](http://www.opensource.org)).
>    **“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
>    Project, including but not limited to communication on electronic mailing lists, source code control
>    systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
>    discussing and improving that Project, but excluding communication that is conspicuously marked or
>    otherwise designated in writing by You as “Not a Submission.”
>    **“Submission”** means the Code and any other copyrightable material Submitted by You, including any
>    associated comments and documentation.
> 2. **Your Submission**. You must agree to the terms of this Agreement before making a Submission to any
>    Project. This Agreement covers any and all Submissions that You, now or in the future (except as
>    described in Section 4 below), Submit to any Project.
> 3. **Originality of Work**. You represent that each of Your Submissions is entirely Your original work.
>    Should You wish to Submit materials that are not Your original work, You may Submit them separately
>    to the Project if You (a) retain all copyright and license information that was in the materials as You
>    received them, (b) in the description accompanying Your Submission, include the phrase “Submission
>    containing materials of a third party:” followed by the names of the third party and any licenses or other
>    restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
>    guidelines concerning Submissions.
> 4. **Your Employer**. References to “employer” in this Agreement include Your employer or anyone else
>    for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
>    Submission is made in the course of Your work for an employer or Your employer has intellectual
>    property rights in Your Submission by contract or applicable law, You must secure permission from Your
>    employer to make the Submission before signing this Agreement. In that case, the term “You” in this
>    Agreement will refer to You and the employer collectively. If You change employers in the future and
>    desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
>    and secure permission from the new employer before Submitting those Submissions.
> 5. **Licenses**.
> 
> * **Copyright License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
>   Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
>   the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
>   parties.
> * **Patent License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
>   Your patent claims that are necessarily infringed by the Submission or the combination of the
>   Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
>   import or otherwise dispose of the Submission alone or with the Project.
> * **Other Rights Reserved**. Each party reserves all rights not expressly granted in this Agreement.
>   No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
>   granted by implication, exhaustion, estoppel or otherwise.
> 
> 6. **Representations and Warranties**. You represent that You are legally entitled to grant the above
>    licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
>    have disclosed under Section 3). You represent that You have secured permission from Your employer to
>    make the Submission in cases where Your Submission is made in the course of Your work for Your
>    employer or Your employer has intellectual property rights in Your Submission by contract or applicable
>    law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
>    have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
>    You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
>    REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
>    EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
>    PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
>    NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
> 7. **Notice to Microsoft**. You agree to notify Microsoft in writing of any facts or circumstances of which
>    You later become aware that would make Your representations in this Agreement inaccurate in any
>    respect.
> 8. **Information about Submissions**. You agree that contributions to Projects and information about
>    contributions may be maintained indefinitely and disclosed publicly, including Your name and other
>    information that You submit with Your Submission.
> 9. **Governing Law/Jurisdiction**. This Agreement is governed by the laws of the State of Washington, and
>    the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
>    Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
>    exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
>    defenses of lack of personal jurisdiction and forum non-conveniens.
> 10. **Entire Agreement/Assignment**. This Agreement is the entire agreement between the parties, and
>     supersedes any and all prior agreements, understandings or communications, written or oral, between
>     the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.

@microsoft-github-policy-service agree

**Comment by WaelKarkoub** [comment]
@sonichi let me know what you think of this approach

**Comment by sonichi** [comment]
I don't know the downstream use case, so I don't know if name conflict is an issue. Presumably it is. One way to handle it is to add a bool "raise_on_name_conflict". The desired structure of it also depends on the downstream use case.

I recommend moving the public methods into `GroupChat`. We don't recommend adding public methods in Agents because it hinders composability.
cc @ekzhu

**Comment by ekzhu** [comment]
Thanks for the PR! Could you share your usage scenario that requires this feature? I am curious why you need to access all agents from a specific team? What's the operation you are performing on these agents?

**Comment by Tylersuard** [comment]
I'm not sure if this is relevant to this discussion or @WaelKarkoub 's PR, but a paper came out saying that the more agents used for a task, the better the performance, as long as the result is voted on by the agents.

File: autogen/agentchat/groupchat.py [hunk 1]
```diff
@@ -478,6 +478,22 @@ def __init__(
...
+        return self.all_agents().get(name)
+
+    def all_agents(self) -> Dict[str, Agent]:
+        """Returns all agents in the group chat manager."""
+        agents = dict()
+        for agent in self._groupchat.agents:
+            if not isinstance(agent, GroupChatManager):
+                agents[agent.name] = agent
+            else:
+                agents.update(agent.all_agents())
```
Comment (joshkyh) [inline autogen/agentchat/groupchat.py:494]: Please consider comment # Recursive call for nested team

File: test/agentchat/test_groupchat.py [hunk 2]
```diff
@@ -672,6 +672,38 @@ def test_clear_agents_history():
...
+        member1 = agent(f"member1_{name}")
+        member2 = agent(f"member2_{name}")
+
+        gc = autogen.GroupChat(agents=[member1, member2], messages=[])
+
+        return autogen.GroupChatManager(groupchat=gc, name=name, llm_config=False)
+
+    user = agent("user")
+    team1 = team("team1")
+    team2 = team("team2")
```
Comment (joshkyh) [inline test/agentchat/test_groupchat.py:694]: Can we consider a new test case of nesting team2B under team2 and see that nested group chat works, please?

**Review by joshkyh — COMMENTED** [review]
Thanks for the PR, excited about the prospect of team operations.

**Comment by afourney** [comment]
Quick question: If we just made the group_chat reference public in GroupChatManager, would that be sufficient? https://github.com/microsoft/autogen/blob/d01063d23122ae2a4bbaa3014519d6ae75e95de1/autogen/agentchat/groupchat.py#L467C33-L467C36

You can get all the agents from there.

File: test/agentchat/test_groupchat.py (hunk 2 above)
Comment (WaelKarkoub) [reply test/agentchat/test_groupchat.py:695]: I've added a test for chat functionality between agents within the main shared group chat, but managed by different group chat managers. Let me know what you think.

**Comment by WaelKarkoub** [comment]
@afourney yes that would solve most of the problems, I've made some changes to the PR where the implementation is moved to the `GroupChat` class. I had to make `_groupchat` in `GroupChatManager` public. I thought implementing `all_agents` would save people time from implementing their own utils.

**Review by joshkyh — APPROVED** [review]
Yeap, thanks! I see that there are now two layers of gc. LGTM

**Comment by sonichi** [comment]
> @afourney yes that would solve most of the problems, I've made some changes to the PR where the implementation is moved to the `GroupChat` class. I had to make `_groupchat` in `GroupChatManager` public. I thought implementing `all_agents` would save people time from implementing their own utils.

This change will affect workflowmanager.py in autogenstudio which used `_groupchat`. @victordibia

File: autogen/agentchat/groupchat.py [hunk 3]
```diff
@@ -178,6 +178,18 @@ def agent_by_name(self, name: str) -> Agent:
         """Returns the agent with a given name."""
         return self.agents[self.agent_names.index(name)]
 
+    def all_agents(self) -> Dict[str, Agent]:
```
Comment (davorrunje) [inline autogen/agentchat/groupchat.py:181]: As far as I see, there is no guarantee that agent names are unique. However, this API assumes they are. In case of collision (two distinct agents having the same name), this function will not return both of them. It would be better to return a list of agents.

File: autogen/agentchat/groupchat.py [hunk 4]
```diff
@@ -464,7 +476,7 @@ def __init__(
             **kwargs,
         )
         # Store groupchat
-        self._groupchat = groupchat
+        self.groupchat = groupchat
```
Comment (davorrunje) [inline autogen/agentchat/groupchat.py:479]: why this change?

Comment (WaelKarkoub) [reply autogen/agentchat/groupchat.py:486]: For nested agents, the only way to get access to the groupchat is through `_groupchat`, but I can't access it since it's a private variable. From my understanding, autogenstudio also needs access to the groupchat, so I thought it might be cleaner to change it to a public variable.

Comment (davorrunje) [reply autogen/agentchat/groupchat.py:488]: The cleanest way would be to define a new property `groupchat`, you don't want this variable to be changed from the outside.

File: autogen/agentchat/groupchat.py [hunk 5]
```diff
@@ -174,9 +175,29 @@ def append(self, message: Dict, speaker: Agent):
...
+        self, name: str, recursive: bool = False, raise_on_name_conflict: bool = False
+    ) -> Optional[Agent]:
+        """Returns the agent with a given name. If recursive is True, it will search in nested teams."""
+        agents = self.nested_agents() if recursive else self.agents
+        filtered_agents = [agent for agent in agents if agent.name == name]
+
+        if raise_on_name_conflict and len(filtered_agents) > 1:
+            raise AgentNameConflict()
+
+        return filtered_agents[0]
```
Comment (davorrunje) [inline autogen/agentchat/groupchat.py:188]: if `filtered_agents` is an empty list, this will raise an exception instead of returning `None`.

File: autogen/agentchat/groupchat.py [hunk 6]
```diff
@@ -174,9 +175,29 @@ def append(self, message: Dict, speaker: Agent):
         message["content"] = content_str(message["content"])
         self.messages.append(message)
 
-    def agent_by_name(self, name: str) -> Agent:
-        """Returns the agent with a given name."""
-        return self.agents[self.agent_names.index(name)]
+    def agent_by_name(
```
Comment (davorrunje) [inline autogen/agentchat/groupchat.py:178]: You should add a test for this function and make sure you test all edge cases with 100% coverage. It is too easy to make a simple mistake that will be hard to catch later.

File: autogen/agentchat/groupchat.py [hunk 7]
```diff
@@ -174,9 +175,29 @@ def append(self, message: Dict, speaker: Agent):
...
+        """Returns all agents in the group chat manager."""
+        agents = list()
+        for agent in self.agents:
+            if not isinstance(agent, GroupChatManager):
+                agents.append(agent)
+            else:
+                agents.append(agent)
+                # Recursive call for nested teams
+                agents.extend(agent.groupchat.nested_agents())
+        return agents
```
Comment (davorrunje) [inline autogen/agentchat/groupchat.py:200]:
```suggestion
        agents = self.agents.copy()
        for agent in agents:
            if isinstance(agent, GroupChatManager):
                # Recursive call for nested teams
                agents.extend(agent.groupchat.nested_agents())
        return agents
```

File: autogen/agentchat/groupchat.py (hunk 6 above)
Comment (WaelKarkoub) [reply autogen/agentchat/groupchat.py:178]: Tests added!

Comment (davorrunje) [reply autogen/agentchat/groupchat.py:178]: what is the test coverage? are there any lines missed?

File: autogen/agentchat/groupchat.py [hunk 8]
```diff
@@ -174,9 +175,27 @@ def append(self, message: Dict, speaker: Agent):
...
+    ) -> Optional[Agent]:
+        """Returns the agent with a given name. If recursive is True, it will search in nested teams."""
+        agents = self.nested_agents() if recursive else self.agents
+        filtered_agents = [agent for agent in agents if agent.name == name]
+
+        if raise_on_name_conflict and len(filtered_agents) > 1:
+            raise AgentNameConflict()
+
+        if filtered_agents:
+            return filtered_agents[0]
```
Comment (davorrunje) [inline autogen/agentchat/groupchat.py:189]:
```suggestion
        return filtered_agents[0] if filtered_agents else None
```
